docs/docs/typelevel.md

```diff
@@ -0,0 +1,449 @@
+ - All operations on array elements forward to generic operations in class `Numeric`.
+
+It would be quite hard for even a good optimizer to undo the generic abstractions and replace them with something simpler if `N` is specialized to a primitive type like `Double`. But since `apply` is marked `transparent`, the specialization happens automatically as a result of inlining the body of `apply` with the new types. Indeed, the specialized version of `dotProduct` looks like this:
+```
+def dotProduct(xs: Array[Double], ys: Array[Double]): Double = {
+  require(xs.length == ys.length)
+  var i = 0
+  var s: Double = n.zero
+  while (i < xs.length) {
+    s = s + xs(i) * ys(i)
+    i += 1
+  }
+  s
+}
+```
+In other words, specialization with transparent functions allows "abstraction without regret". The price to pay for this
+is the increase of code size through inlining. That price is often worth paying, but inlining decisions need to be considered carefully.
+
+## Implicit Matches
+
+It is foreseen that many areas of typelevel programming can be done with transparent functions instead of implicits. But sometimes implicits are unavoidable. The problem so far was that the Prolog-like programming style of implicit search becomes viral: Once some construct depends on implicit search it has to be written as a logic program itself. Consider for instance the problem
+of creating a `TreeSet[T]` or a `HashSet[T]` depending on whether `T` has an `Ordering` or not. We can create a set of implicit definitions like this:
+```scala
+trait SetFor[T, S <: Set[T]]
+class LowPriority {
+  implicit def hashSetFor[T]: SetFor[T, HashSet[T]] = ...
+}
+object SetsFor extends LowPriority {
+  implicit def treeSetFor[T: Ordering]: SetFor[T, TreeSet[T]] = ...
+}
+```
+Clearly, this is not pretty. Besides all the usual indirection of implicit search,
+we face the problem of rule prioritization where
+we have to ensure that `treeSetFor` takes priority over `hashSetFor` if the element type has an ordering. This is solved
+(clumsily) by putting `hashSetFor` in a superclass `LowPriority` of the object `SetsFor` where `treeSetFor` is defined.
+Maybe the boilerplate would still be acceptable if the crufty code could be contained. However, this is not the case. Every user of
+the abstraction has to be parameterized itself with a `SetFor` implicit. Considering the simple task _"I want a `TreeSet[T]` if
+`T` has an ordering and a `HashSet[T]` otherwise"_, this seems like a lot of ceremony.
+
+There are some proposals to improve the situation in specific areas, for instance by allowing more elaborate schemes to specify priorities. But they all keep the viral nature of implicit search programs based on logic programming.
+
+By contrast, the new `implicit match` construct makes implicit search available in a functional context. To solve
+the problem of creating the right set, one would use it as follows:
+```scala
+transparent def setFor[T]: Set[T] = implicit match {
```

Comment:
This seems very ad hoc. It seems like there's a lot of potential for generalization.
By the way, would `setFor` typically be marked `implicit` itself?
Anyway, couldn't this somehow be written without `implicit match` if we had some (`transparent`) way to check whether an implicit exists? I would start by thinking of how to compose that with existing code.
Something like
```scala
transparent maybeImplicit[T](implicit t: T = null): Option[T] = Option(t)
transparent def setFor[T] =
  maybeImplicit[Ordering[T]] match {
    case None => new HashSet[T]
    case Some(ordering) =>
      implicit val o = ordering
      new TreeSet[T]
  }
```

So, it seems to me that such a construct is possible without introducing `implicit match`. Perhaps it would force `transparent` unrolling to work a little harder. Also, it would be nice if we would allow the `implicit` modifier on pattern bindings -- then you could do `case Some(implicit ordering) =>` (or even `case Some(implicit _) => `!) But that would be useful generally too. Also, I don't know if `transparent` can unroll HOFs, but if it could then the above could be written with map/getOrElse or with fold.

That said, pattern matching on types seems like it could be a really nice feature, but again, I would hope it could be done in a general way, not tied to implicit search. At least, implicit search shouldn't "squat" on syntax that could be left open to the potential of such a feature.

Reply:
For this to work, `maybeImplicit` has to be "magic". In the end we need some way to try to summon an implicit without causing a compile-time error if none is found. So, given that some new compiler-supported machinery is needed, `implicit match` is a more pleasant to use than `maybeImplicit`.

Reply:
@odersky in which ways does `maybeImplicit` need to be magic?

FWIW, the following currently works:

```scala
scala> def maybeImplicit[T](implicit t: T = null): Option[T] = Option(t)
maybeImplicit: [T](implicit t: T)Option[T]

scala> maybeImplicit[String]
res0: Option[String] = None

scala> implicit val str = "ok"
str: String = ok

scala> maybeImplicit[String]
res4: Option[String] = Some(ok)
```